llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp

```diff
@@ -0,0 +1,186 @@
+      // get the size and offset node to calculate the total size of spec
+      // constants (size of type + offset in the composite - if any);
+      PerKernelSize += uintFromMDNode(SC, SC->getNumOperands() - 1);
+      PerKernelSize += uintFromMDNode(SC, SC->getNumOperands() - 2);
+    }
+    // finally, allocate per-kernel globals.
+    std::string GlobalName{GlobalNamePrefix};
+    GlobalName.append(KernelName.str());
+    Globals[KernelName] = createGlobal(M, PerKernelSize, GlobalName);
+  }
+
+  auto &Context = M.getContext();
+  IRBuilder B(Context);
+  SmallVector<Instruction *> ToErase;
+  // Now, re-write the signature of the kernel to drop the extra spec constant
+  // argument and patch in accesses to the global symbol (instead of the arg).
+  for (const auto *Node : MD->operands()) {
+    const StringRef KernelName =
+        cast<MDString>(Node->getOperand(0).get())->getString();
+    auto *F = M.getFunction(KernelName);
+    assert(F && "Function not found.");
+    auto *SpecConstArg = std::prev(F->arg_end());
+    assert(SpecConstArg->getName() == "_arg__specialization_constants_buffer" &&
+           "Expected a kernel.");
+
+    // Prepare a new function type, a copy of the original without the last arg.
+    std::vector<Type *> Arguments;
+    SmallVector<AttributeSet, 8> ArgumentAttributes;
+    auto FAttrs =
+        F->getAttributes().removeParamAttributes(Context, F->arg_size() - 1);
+    for (unsigned i = 0; i < F->arg_size() - 1; ++i) {
+      Arguments.push_back(F->getArg(i)->getType());
+      ArgumentAttributes.push_back(FAttrs.getParamAttrs(i));
+    }
+    AttributeList NAttrs = AttributeList::get(
+        Context, FAttrs.getFnAttrs(), FAttrs.getRetAttrs(), ArgumentAttributes);
+    FunctionType *NFTy =
+        FunctionType::get(F->getFunctionType()->getReturnType(), Arguments,
+                          F->getFunctionType()->isVarArg());
+    // Create a new function body and insert it into the module.
+    F->setName(KernelName + "_old");
+    Function *NF = Function::Create(NFTy, F->getLinkage(), F->getAddressSpace(),
+                                    KernelName, &M);
+    NF->setComdat(F->getComdat());
+    NF->setDSOLocal(true);
+    NF->setUnnamedAddr(GlobalValue::UnnamedAddr::Local);
+    NF->setAttributes(NAttrs);
+
+    // Re-write uses of, now defunct, spec constant kernel arg to GEPs from the
+    // global value.
+    for (auto *U : SpecConstArg->users()) {
+      auto *I = llvm::dyn_cast_or_null<Instruction>(&*U);
+      assert(I && "Expected an instruction.");
+      switch (I->getOpcode()) {
+      default: {
+        std::string Str;
+        llvm::raw_string_ostream Out(Str);
+        Out << "Unhandled instruction: ";
+        I->print(Out);
+        llvm_unreachable(Str.c_str());
+      }
+      case Instruction::AddrSpaceCast: {
+        auto *ASCInst = dyn_cast_or_null<AddrSpaceCastInst>(I);
+        // TODO: getPointerElementType will have to be changed once opaque
```

Comment:
You should just use [`PointerType::withSamePointeeType`](https://llvm.org/doxygen/classllvm_1_1PointerType.html#ac0f741735997264d3018e30ce2a6b2d3) right away

Reply:
Done